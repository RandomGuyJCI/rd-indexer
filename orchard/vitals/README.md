# vitals

This is a script which extracts metadata from a rdzip.

# usage

`poetry run python ./orchard/vitals/vitals.py <path to rdzip>`

