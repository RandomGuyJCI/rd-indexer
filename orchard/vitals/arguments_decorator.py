import functools


def with_arguments(*argument_names):
    """
    A decorator that we put on facets to let facets declare what arguments they actually use.
    The decorator fetches the required values from the dict and gives them to the function in order.
    """

    def decorator(func):
        @functools.wraps(func)
        def wrapper(obj):
            args = [obj[k] for k in argument_names]
            return func(*args)

        return wrapper

    return decorator
