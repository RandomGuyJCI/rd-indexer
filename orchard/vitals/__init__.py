from orchard.vitals.vitals import main as analyze