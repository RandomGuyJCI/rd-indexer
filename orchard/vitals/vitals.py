import sys
import zipfile
from typing import BinaryIO
from orchard.parse import parse
from orchard.vitals.facets.author_facet import author_facet
from orchard.vitals.facets.beat_types_facet import beat_type_facet
from orchard.vitals.facets.bpm_facet import bpm_facet
from orchard.vitals.facets.difficulty_facet import difficulty_facet
from orchard.vitals.facets.icon_facet import icon_facet
from orchard.vitals.facets.id_facet import id_facet
from orchard.vitals.facets.key_facet import make_key_facet
from orchard.vitals.facets.player_facet import player_facet
from orchard.vitals.facets.tags_facet import tags_facet
from orchard.vitals.facets.thumbnail_facet import thumbnail_facet
from orchard.vitals.facets.updated_facet import updated_facet


class VitalsException(Exception):
    pass


def main(f: BinaryIO):
    facets = {
        "id":                               id_facet,
        "artist":                           make_key_facet(["settings", "artist"]),
        "song":                             make_key_facet(["settings", "song"]),
        "seizure_warning":                  make_key_facet(["settings", "seizureWarning"], True),
        "description":                      make_key_facet(["settings", "description"]),
        "hue":                              make_key_facet(["settings", "songNameHue"], 0.0),
        "authors":                          author_facet,
        ("max_bpm", "min_bpm"):             bpm_facet,
        "difficulty":                       difficulty_facet,
        ("single_player", "two_player"):    player_facet,
        "last_updated":                     updated_facet,
        "tags":                             tags_facet,
        ("image", "thumb"):                 thumbnail_facet,
        "icon":                             icon_facet,
        (
            "has_classics",
            "has_oneshots",
            "has_squareshots",
            "has_freezeshots",
            "has_freetimes",
            "has_holds"
        ):                                  beat_type_facet
    }

    try:
        with zipfile.ZipFile(f) as z:
            with z.open('main.rdlevel', 'r') as rdlevel:
                text = rdlevel.read().decode('utf-8-sig')
                parsed = parse(text)

                final = {}
                for key, func in facets.items():
                    try:
                        result = func(parsed, z)
                        if isinstance(key, tuple):
                            # multiple keys with (expected) multiple values that map to the keys.
                            for k, v in zip(key, result):
                                final[k] = v
                        else:
                            final[key] = result
                    except Exception as e:
                        raise VitalsException(f"vitals: An unhandled error occured in a facet: {e}")

                return final

    except zipfile.BadZipFile:
        raise VitalsException("vitals: this is not a zip file, or we couldn't decode it for some reason.")
    except KeyError:
        raise VitalsException("vitals: there is no main.rdlevel in the zip.")

if __name__ == "__main__":
    try:
        file_name = sys.argv[1]
        with open(file_name, 'rb') as f:
            vit = main(f)
    except IndexError:
        print("Did you pass a file name in?")