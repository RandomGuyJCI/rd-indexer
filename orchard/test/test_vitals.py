from orchard.vitals.vitals import main


def test_haha_awesome(rdzip, snapshot):
    with open(rdzip, "rb") as f:
        assert main(f) == snapshot
    assert True