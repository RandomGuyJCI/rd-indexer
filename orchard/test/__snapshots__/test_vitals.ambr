# name: test_haha_awesome[./fixtures\\Andrew_Huang_-_Know_You.rdzip]
  dict({
    'artist': 'Andrew Huang',
    'artist_tokens': list([
      'Andrew Huang',
    ]),
    'authors': list([
      'lugi',
    ]),
    'description': 'Two lonely stars, wandering in the dark...',
    'difficulty': 1,
    'has_classics': True,
    'has_freetimes': True,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': True,
    'has_skipshots': False,
    'has_squareshots': False,
    'has_window_dance': False,
    'hue': 0.47,
    'id': 'know-you-2zRJRCjPqJx',
    'last_updated': datetime.datetime(2020, 6, 24, 15, 6, 48),
    'max_bpm': 78.5,
    'min_bpm': 78.5,
    'seizure_warning': False,
    'sha1': 'f2ad0f76bbfedf96bec3451eb9ead014accb3047',
    'single_player': True,
    'song': 'Know You',
    'tags': list([
      'slow',
      'pop',
    ]),
    'two_player': True,
  })
# ---
# name: test_haha_awesome[./fixtures\\Hot_Freaks_-_Pull.rdzip]
  dict({
    'artist': 'Hot Freaks',
    'artist_tokens': list([
      'Hot Freaks',
    ]),
    'authors': list([
      'Labyrinth',
    ]),
    'description': "It's not something easy to admit.",
    'difficulty': 1,
    'has_classics': True,
    'has_freetimes': True,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': False,
    'has_skipshots': False,
    'has_squareshots': False,
    'has_window_dance': False,
    'hue': 0.88,
    'id': 'pull-EpgEYcGYPnU',
    'last_updated': datetime.datetime(2022, 6, 16, 0, 8, 26),
    'max_bpm': 89,
    'min_bpm': 88,
    'seizure_warning': False,
    'sha1': '229e4a42f2e69db34d851fd6d66618c58ca076e4',
    'single_player': True,
    'song': 'Pull',
    'tags': list([
      '1p',
      '2p',
      'gimmick',
    ]),
    'two_player': True,
  })
# ---
# name: test_haha_awesome[./fixtures\\Pseuu_-_ennui.rdzip]
  dict({
    'artist': 'Pseuu',
    'artist_tokens': list([
      'Pseuu',
    ]),
    'authors': list([
      'Pseuu',
    ]),
    'description': '''
      uh... no comment. level made in about 3-4 working hours
      
      - tricky oneshot patterns
      - skipshots
      - holds
    ''',
    'difficulty': 2,
    'has_classics': False,
    'has_freetimes': True,
    'has_freezeshots': False,
    'has_holds': True,
    'has_oneshots': True,
    'has_skipshots': True,
    'has_squareshots': False,
    'has_window_dance': True,
    'hue': 0.42,
    'id': 'ennui-og-3VFE7LSmzSo',
    'last_updated': datetime.datetime(2022, 1, 3, 23, 19, 26),
    'max_bpm': 126,
    'min_bpm': 126,
    'seizure_warning': False,
    'sha1': '3031e6cd0842aedec80d9a0dbd8866f25cc58ec2',
    'single_player': True,
    'song': 'ennui.ogg',
    'tags': list([
      'chiptune',
      'short',
      '2p',
      'oneshots',
      'skipshots',
      'holds',
    ]),
    'two_player': True,
  })
# ---
# name: test_haha_awesome[./fixtures\\Quree_-_One_Forgotten_Night.rdzip]
  dict({
    'artist': 'Quree',
    'artist_tokens': list([
      'Quree',
    ]),
    'authors': list([
      'BluPen',
    ]),
    'description': "Patient experiencing sudden heart strain again. Solve the patient's concerns.",
    'difficulty': 2,
    'has_classics': True,
    'has_freetimes': False,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': True,
    'has_skipshots': False,
    'has_squareshots': False,
    'has_window_dance': True,
    'hue': 0.1,
    'id': 'one-forg-LJfg3wQANR7',
    'last_updated': datetime.datetime(2021, 5, 20, 21, 9, 22),
    'max_bpm': 173,
    'min_bpm': 100,
    'seizure_warning': False,
    'sha1': '68f8771d0898813755bdc26cb6c79bd822824cb5',
    'single_player': True,
    'song': 'One Forgotten Night',
    'tags': list([
      'quree',
      'window dance',
    ]),
    'two_player': True,
  })
# ---
# name: test_haha_awesome[./fixtures\\Samario (Making A Mistake) - One Uncued Tresillo.rdzip]
  dict({
    'artist': 'Samario (Making A Mistake)',
    'artist_tokens': list([
      'Samario',
    ]),
    'authors': list([
      'Samario (Making A Mistake)',
    ]),
    'description': '''
      One Uncued Tresillo
      (song edited from [...]Machinegun Psystyle[...], by Camellia)
    ''',
    'difficulty': 0,
    'has_classics': False,
    'has_freetimes': True,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': False,
    'has_skipshots': False,
    'has_squareshots': False,
    'has_window_dance': False,
    'hue': 0,
    'id': 'one-uncu-i64TVVfB93P',
    'last_updated': datetime.datetime(2022, 6, 27, 21, 27, 12),
    'max_bpm': 70,
    'min_bpm': 40,
    'seizure_warning': False,
    'sha1': '024b74e8a3ce6ae1215b49ec2025cd75e0a6b468',
    'single_player': True,
    'song': 'One Uncued Tresillo',
    'tags': list([
      'uncued tresillo',
      'samario please stop doing this',
    ]),
    'two_player': True,
  })
# ---
# name: test_haha_awesome[./fixtures\\Samario_Making_A_Mistake_-_One_Uncued_Tresillo orig.rdzip]
  dict({
    'artist': 'Samario (Making A Mistake)',
    'artist_tokens': list([
      'Samario (Making A Mistake)',
    ]),
    'authors': list([
      'Samario (Making A Mistake)',
    ]),
    'description': '''
      One Uncued Tresillo
      (song edited from [...]Machinegun Psystyle[...], by Camellia)
    ''',
    'difficulty': 0,
    'has_classics': False,
    'has_freetimes': True,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': False,
    'has_skipshots': False,
    'has_squareshots': False,
    'has_window_dance': False,
    'hue': 0,
    'id': 'one-uncu-eqSkB1ZX7Xz',
    'last_updated': datetime.datetime(2021, 1, 2, 3, 52, 10),
    'max_bpm': 160,
    'min_bpm': 160,
    'seizure_warning': False,
    'sha1': 'd7fb093e54173c553a9603929df7267b27267d1f',
    'single_player': True,
    'song': 'One Uncued Tresillo',
    'tags': list([
      'uncued tresillo',
      'samario please stop doing this',
    ]),
    'two_player': True,
  })
# ---
# name: test_haha_awesome[./fixtures\\auburnsummer - Triplet Testaaaaa.rdzip]
  dict({
    'artist': 'auburnsummer',
    'artist_tokens': list([
      'auburnsummer',
    ]),
    'authors': list([
      'auburnsummer',
      'Xeno',
    ]),
    'description': 'A two-handed level. The eloquent blend between thirds and halves.',
    'difficulty': 2,
    'has_classics': True,
    'has_freetimes': False,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': True,
    'has_skipshots': False,
    'has_squareshots': True,
    'has_window_dance': False,
    'hue': 0.66,
    'id': 'triplet-LrxHzPMLamy',
    'last_updated': datetime.datetime(2022, 6, 27, 23, 25, 28),
    'max_bpm': 110,
    'min_bpm': 110,
    'seizure_warning': False,
    'sha1': '4de04dc0c7e6ba30438368cacbd8a24678666841',
    'single_player': True,
    'song': 'Triplet Test',
    'tags': list([
      'Piano',
      '2p',
      'Two-handed',
    ]),
    'two_player': False,
  })
# ---
# name: test_haha_awesome[./fixtures\\placeboing_-_BANAS.rdzip]
  dict({
    'artist': 'placeboing',
    'artist_tokens': list([
      'placeboing',
    ]),
    'authors': list([
      "April's Only Fool",
    ]),
    'description': '''
      This level works under the assumption that you know how 8beat, 6beat, and Temporal Shift works.
      
      Hope you all have a wonderful April Fools Day!
      
      Hollaback Girl by Gwen Stefani x Donkey Kong Country 2 OST
      
      Banana CC by Pyre
      Banana Tile by Het
    ''',
    'difficulty': 1,
    'has_classics': True,
    'has_freetimes': True,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': True,
    'has_skipshots': True,
    'has_squareshots': False,
    'has_window_dance': False,
    'hue': 0.17,
    'id': 'banas-QbMoyaG2nfK',
    'last_updated': datetime.datetime(2022, 4, 9, 16, 40, 34),
    'max_bpm': 147,
    'min_bpm': 143.5,
    'seizure_warning': False,
    'sha1': '053aac4ed9987f1a6e8bdf12df120a9d607879f7',
    'single_player': True,
    'song': 'BANAS',
    'tags': list([
      'April Fools',
      'banana',
      'banas',
      'temporal shift',
      '8beat',
      '6beat',
    ]),
    'two_player': False,
  })
# ---
# name: test_haha_awesome[./fixtures\\toonlink_Kenichi_Nishikami_Takeru_Kanazaki_-_Jimmys_Aesthetics.rdzip]
  dict({
    'artist': 'toonlink, Kenichi Nishikami, Takeru Kanazaki',
    'artist_tokens': list([
      'toonlink',
      'Kenichi Nishikami',
      'Takeru Kanazaki',
    ]),
    'authors': list([
      'Szprycha',
    ]),
    'description': 'A neat song from WarioWare DIY',
    'difficulty': 1,
    'has_classics': True,
    'has_freetimes': False,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': True,
    'has_skipshots': False,
    'has_squareshots': False,
    'has_window_dance': False,
    'hue': 0.19,
    'id': 'jimmy-s-Ccby1RK3XJK',
    'last_updated': datetime.datetime(2020, 6, 20, 23, 9, 28),
    'max_bpm': 125,
    'min_bpm': 125,
    'seizure_warning': False,
    'sha1': '8da77730d1fcd96d73590505b2a502d0dd05807c',
    'single_player': True,
    'song': "Jimmy's Aesthetics",
    'tags': list([
      'siivagunner',
      'nintendo',
      'warioware',
      'body rock',
    ]),
    'two_player': False,
  })
# ---
